src/arithmetic/arithmetic_expression.c

```diff
@@ -1178,6 +1149,169 @@ SIValue AR_NE(SIValue *argv, int argc) {
+	// if index out of bound, end at arrayLen
+	endIndex = ((int32_t)arrayLen) < endIndex ? arrayLen : endIndex;
+	// cant go in reverse
+	if(endIndex <= startIndex) {
+		return SI_EmptyArray();
+	}
+
+	SIValue subArray = SI_Array(endIndex - startIndex);
+	for(uint i = startIndex; i < endIndex; i++) {
+		SIArray_Append(&subArray, SIArray_Get(array, i));
+	}
+	return subArray;
+}
+
+// create a new list of integers in the range of [start, end]. If an interval was given
+// the interval between two consecutive list members will be this interval.
+// If interval was not suppllied, it will be default as 1
+// "RETURN range(3,8,2)" will yield [3, 5, 7]
+SIValue AR_RANGE(SIValue *argv, int argc) {
+	assert(argc > 1 && argc <= 3 && argv[0].type == T_INT64 && argv[1].type == T_INT64);
+	int64_t start = argv[0].longval;
+	int64_t end = argv[1].longval;
+	int64_t interval = 1;
+	if(argc == 3) {
+		assert(argv[2].type == T_INT64);
+		interval = argv[2].longval;
+	}
+	SIValue array = SI_Array(1 + (end - start) / interval);
+	for(; start <= end; start += interval) {
+		SIArray_Append(&array, SI_LongVal(start));
+	}
+	return array;
+}
+
+// checks if a value is in a given list
+// "RETURN 3 IN [1, 2, 3]" will return true
+SIValue AR_IN(SIValue *argv, int argc) {
+	assert(argc == 2 && argv[1].type == T_ARRAY);
+	SIValue lookupValue = argv[0];
+	SIValue lookupList = argv[1];
+	bool comparedNull = false;
+	uint arrayLen = SIArray_Length(lookupList);
+	for(uint i = 0; i < arrayLen; i++) {
+		int compareValue = SIValue_Compare(lookupValue, SIArray_Get(lookupList, i));
+		if(compareValue == 0) return SI_BoolVal(true);
+		if(compareValue == COMPARED_NULL) comparedNull = true;
+	}
+	return comparedNull ? SI_NullVal() : SI_BoolVal(false);
+}
+
+// return a list/string/map/path size
+// "RETURN size([1, 2, 3])" will return 3
+// TODO: when map and path are implemented, add their functonality
+SIValue AR_SIZE(SIValue *argv, int argc) {
+	assert(argc == 1);
+	SIValue value = argv[0];
+	switch(value.type) {
+	case T_ARRAY:
+		return SI_LongVal(SIArray_Length(value));
+	case T_STRING:
+		return SI_LongVal(strlen(value.stringval));
+	default:
+		assert(false);
+	}
+}
+
+// return the first member of a list
+// "RETURN head([1, 2, 3])" will return 1
+SIValue AR_HEAD(SIValue *argv, int argc) {
+	assert(argc == 1);
+	SIValue value = argv[0];
+	if(value.type == T_NULL) return SI_NullVal();
+	assert(value.type == T_ARRAY);
+	uint arrayLen = SIArray_Length(value);
+	if(arrayLen == 0) return SI_NullVal();
+	return SIArray_Get(value, 0);
+}
+
+// return a sublist of a list, which contains all the values withiout the first value
+// "RETURN tail([1, 2, 3])" will return [2, 3]
+SIValue AR_TAIL(SIValue *argv, int argc) {
+	assert(argc == 1);
+	SIValue value = argv[0];
+	if(value.type == T_NULL) return SI_NullVal();
+	assert(value.type == T_ARRAY);
+	uint arrayLen = SIArray_Length(value);
+	SIValue array = SI_Array(arrayLen);
+	if(arrayLen < 2) return array;
```

Comment:
You likely need to clone here or do something with the `SIAllocation`, as in all other cases this call will allocate memory.

Reply:
what is wrong with this memory allocation? a new array value is requested by the caller